clang/lib/Driver/ToolChains/AMDGPU.cpp

```diff
@@ -923,6 +923,12 @@ ROCMToolChain::getCommonDeviceLibNames(const llvm::opt::ArgList &DriverArgs,
   bool CorrectSqrt = DriverArgs.hasFlag(
       options::OPT_fhip_fp32_correctly_rounded_divide_sqrt,
       options::OPT_fno_hip_fp32_correctly_rounded_divide_sqrt);
+
+  if (DeviceOffloadingKind == Action::OFK_SYCL) {
+    // When using SYCL, sqrt is only correctly rounded if the flag is specified
+    CorrectSqrt = DriverArgs.hasArg(options::OPT_fsycl_fp32_prec_sqrt);
+  }
```

Comment:
Should some sort of an error diagnostic be issued for a conflict between these two settings? I realize that combining `-fsycl-fp32-prec-sqrt` with a contradicting target-specific option would be a rather stupid thing to do, but the UI still has to account for it. Or is such a situation theoretically impossible?
Maybe an even better option, instead of an error-upon-conflict, would be to _make_ it theoretically impossible by placing the target-specific option check inside of an `else` block:
```suggestion
  bool CorrectSqrt = false;
  if (DeviceOffloadingKind == Action::OFK_SYCL) {
    // When using SYCL, sqrt is only correctly rounded if the flag is specified
    CorrectSqrt = DriverArgs.hasArg(options::OPT_fsycl_fp32_prec_sqrt);
  } else
    CorrectSqrt = DriverArgs.hasFlag(
                                 options::OPT_fhip_fp32_correctly_rounded_divide_sqrt,
                                 options::OPT_fno_hip_fp32_correctly_rounded_divide_sqrt);
```
(the long-winded comment applies to both HIP and CUDA paths)

Reply:
Thanks for the explanation - I went a little bit astray when saying CUDA had the same problems. That part can be disregarded in view of the CC1 option specifics.
> I wonder if there's a way to disable that in the argument parsing to get a proper diagnostic, I'll try to see if I can find something.

I believe that an explicit `else` block would have this exact effect - the `-fsycl-fp32-prec-sqrt` flag simply wouldn't  get claimed and the user would get the same "argument unused" warning.

Reply:
As you noticed I applied the `else` change, but I also just extended the `amdgcn` test to check that the warning is printed out when the hip flag is used in SYCL mode.